server/service/src/sensor/update.rs

```diff
@@ -54,6 +60,98 @@ pub fn update_sensor(
+                // use the start of the day as first log time
+                let mut first_breach_datetime = NaiveDateTime::new(first_log.temperature_log_row.datetime.date(), zero_time);
+                let first_log_datetime = first_breach_datetime + Duration::seconds(log_interval.into());
+
+                if first_log_datetime < first_log.temperature_log_row.datetime {
+                    first_breach_datetime = first_log.temperature_log_row.datetime;
+                }
+
+                // Use the first log time if it's either earlier than the calculated breach end
+                // or if the calculated breach start is less than the sensor log interval before the first log time
+                if (breach_record.start_datetime > first_breach_datetime) | (log_interval > 0) & (breach_record.start_datetime < first_breach_datetime - Duration::seconds(log_interval.into())) {
+                    log::info!(
+                        "Updating cumulative breach start for {:?} to {:?}",
+                        breach_record,
+                        first_breach_datetime
+                    );
+                    breach_record.start_datetime = first_breach_datetime;
+                    TemperatureBreachRowRepository::new(connection).upsert_one(&breach_record)?;
+                }
+            }
+            if let Some(last_log) = logs.last() {
+                // If within log_interval seconds of the end of the day,
+                // use the end of the day as last log time
+                let mut last_breach_datetime = NaiveDateTime::new(last_log.temperature_log_row.datetime.date(), zero_time) + Duration::days(1);
+                let last_log_datetime = last_breach_datetime - Duration::seconds(log_interval.into());
+
+                if last_log_datetime > last_log.temperature_log_row.datetime {
+                    last_breach_datetime = last_log.temperature_log_row.datetime;
+                }
+                
+                // Use the last log time if it's either later than the calculated breach end
+                // or if the calculated breach end is more than the sensor log interval after the last log time
+                if (breach_record.end_datetime < Some(last_breach_datetime)) | (log_interval > 0) & (breach_record.end_datetime > Some(last_breach_datetime + Duration::seconds(log_interval.into()))) {
+                    log::info!(
+                        "Updating cumulative breach end for {:?} to {:?}",
+                        breach_record,
+                        last_breach_datetime
+                    );
+                    breach_record.end_datetime = Some(last_breach_datetime);
+                    TemperatureBreachRowRepository::new(connection).upsert_one(&breach_record)?;
+                }
+            }
+        }
+
+        for mut temperature_log in logs {
+            if let Some(_breach_id) = &temperature_log.temperature_log_row.temperature_breach_id {
```

Comment:
The simple logic is that a re-import starts from the last saved timestamp of the previous import, so any temperature logs or breaches which have finished before that are ignored (and in any case, the import code won't update an existing temperature log or finished breach anyway) => no pre-existing records should re-sync.

Actually it's not quite that simple, as you could have had an ongoing breach at the time of the last saved timestamp - that's why the query for a unique breach doesn't include the end time, so that it will still get updated if the end time has changed.

FYI, the mSupply desktop code takes this a bit further to calculate consecutive breaches from the raw temperature log data - in that case, it rewinds to the end of the latest finished breach before the last saved timestamp (as you might not have had enough breached logs to trigger a breach at that point) => we'll have to do something similar if/when we do that in omSupply

Reply:
> (and in any case, the import code won't update an existing temperature log or finished breach anyway

Do you mean currently or in a simpler logic ? To me, when i read this code, it looks like if there is a consecutive breach, we ignore if breach_id is set or not, and update the logs anyway ? (which would cause a re-sync of logs)

Reply:
It's been that way since the start - no existing records should get updated (or resynced) except for the case where there was an ongoing breach during the last import